src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -0,0 +1,201 @@
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    /// <summary>
+    /// Establishes server endpoint for processes to connect when configured to provide diagnostics connection is reverse mode.
+    /// </summary>
+    internal sealed class ReversedDiagnosticsServer : IDisposable
+    {
+        // The amount of time to allow parsing of the advertise data before cancelling. This allows the server to
+        // remain responsive in case the advertise data is incomplete and the stream is not closed.
+        private static readonly TimeSpan ParseAdvertiseTimeout = TimeSpan.FromSeconds(1);
+
+        private readonly CancellationTokenSource _cancellation = new CancellationTokenSource();
+        private readonly ConcurrentDictionary<Guid, ServerIpcEndpoint> _endpoints = new ConcurrentDictionary<Guid, ServerIpcEndpoint>();
+        private readonly IpcServerTransport _transport;
+
+        private bool _disposed = false;
+
+        /// <summary>
+        /// Constructs the <see cref="ReversedDiagnosticsServer"/> instance with an endpoint bound
+        /// to the location specified by <paramref name="transportPath"/>.
+        /// </summary>
+        /// <param name="transportPath">
+        /// The path of the server endpoint.
+        /// On Windows, this can be a full pipe path or the name without the "\\.\pipe\" prefix.
+        /// On all other systems, this must be the full file path of the socket.
+        /// </param>
+        public ReversedDiagnosticsServer(string transportPath)
+            : this(transportPath, MaxAllowedConnections)
+        {
+        }
+
+        /// <summary>
+        /// Constructs the <see cref="ReversedDiagnosticsServer"/> instance with an endpoint bound
+        /// to the location specified by <paramref name="transportPath"/>.
+        /// </summary>
+        /// <param name="transportPath">
+        /// The path of the server endpoint.
+        /// On Windows, this can be a full pipe path or the name without the "\\.\pipe\" prefix.
+        /// On all other systems, this must be the full file path of the socket.
+        /// </param>
+        /// <param name="maxConnections">The maximum number of connections the server will support.</param>
+        public ReversedDiagnosticsServer(string transportPath, int maxConnections)
+        {
+            _transport = IpcServerTransport.Create(transportPath, maxConnections);
+        }
+
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                _cancellation.Cancel();
+
+                IEnumerable<ServerIpcEndpoint> endpoints = _endpoints.Values;
+                _endpoints.Clear();
+
+                foreach (var endpoint in endpoints)
+                {
+                    endpoint.Dispose();
```

Comment:
Currently I don't think disposing the server triggers any error in pending connection operations for the endpoints. I'm not certain it has to, but I think the convention in similar domains is that it would.

Reply:
I'll look into this an add a test for it too.

Reply:
I've added code to have the pending connections throw ObjectDisposedException when this happens, however, I haven't added a test for it yet. I believe that would largely require that I create a mock client that implements a subset of the diagnostic communication protocol so simulate a client that doesn't reconnect to the server. This mock could be useful in consistently testing other abnormal behaviors and making sure that the server responds appropriately.